src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemIndex.java

```diff
@@ -0,0 +1,182 @@
+package org.broadinstitute.hellbender.utils.bwa;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * BwaMemIndex manages the mapping of a bwa index image file into (non-Java) memory.
+ * It's typically a huge chunk of memory, so you need to manage it as a precious resource.
+ *
+ * Usage pattern is:
+ *   Create a BwaMemIndex for some reference.
+ *   Create Aligners as needed to do some aligning -- they're pretty lightweight and thread safe.
+ *     (But you may need to manage memory by controlling the number of reads that you align in one chunk.)
+ *   Close the BwaMemIndex when you're done aligning.
+ *
+ * This class doesn't know anything about Spark.  You can use it in a distributed setting if you distribute the index
+ * file to each node using the Yarn --files mechanism.  You might find it convenient to manage a singleton instance
+ * of a BwaMemIndex on each Java VM when you're running distributed:  check out BwaMemIndexSingleton in GATK.
+ *
+ * Alternatively, you could use this class directly to run bwa multi-threaded on a single machine,
+ * if that's what you want to do.
+ */
+public final class BwaMemIndex implements AutoCloseable {
+    private final String indexImageFile; // stash this for error messages
+    private volatile long indexAddress; // address where the index was memory-mapped (for use by C code)
+    private final AtomicInteger refCount; // keep track of how many threads are actively aligning
+    private final List<String> refContigNames; // the reference dictionary from the index
+    private static volatile boolean nativeLibLoaded = false; // whether we've loaded the native library or not
+
+    /**
+     * This method creates the new, single-file bwa index image from the 5 files that the "bwa index" command produces.
+     */
+    public static void createIndexImage( final String refName, final String indexImageFile ) {
```

Comment:
I think there should be a test for this. 

Are we going to need a command line tool in gatk to do this?